File: tests/python/stack_overflow.py
#!/usr/bin/env python3
"""Stack overflow via recursion - NO signal expected (usually).

Python has a configurable recursion limit (default 1000). When exceeded,
Python raises RecursionError - a regular Python exception, NOT a signal.
CPython detects this before hitting the actual OS stack limit.

crash-tracer will NOT see this. It's handled entirely in Python-land.

If you set sys.setrecursionlimit() very high, Python might actually
exhaust the OS stack and SIGSEGV, but the default limit prevents that.
"""
import sys

print("[python/stack_overflow] Recursing past Python's recursion limit...", file=sys.stderr)
print(f"[python/stack_overflow] Current limit: {sys.getrecursionlimit()}", file=sys.stderr)

def recurse(n):
    return recurse(n + 1)

try:
    recurse(0)
except RecursionError as e:
    print(f"[python/stack_overflow] Python caught it: {e}", file=sys.stderr)
    print("[python/stack_overflow] No signal generated - process survived", file=sys.stderr)

File: tests/native/stack_overflow.c
// SIGSEGV - stack overflow via infinite recursion
// Expected: signal=11, si_code=SEGV_MAPERR, fault_addr near stack limit
// Note: fault_addr will be near the bottom of the stack mapping, NOT near 0x0
#include <stdio.h>

volatile int depth = 0;

void recurse() {
    char buf[4096];
    buf[0] = (char)depth;
    depth++;
    recurse();
}

int main() {
    fprintf(stderr, "[native/stack_overflow] Recursing until stack exhaustion...\n");
    fflush(stderr);
    recurse();
    return 0;
}
